crates/typst/src/lib.rs

```diff
@@ -402,3 +407,42 @@ fn prelude(global: &mut Scope) {
     global.define("horizon", Alignment::HORIZON);
     global.define("bottom", Alignment::BOTTOM);
 }
+
+/// Adds useful hints when the main source file couldn't be read
+/// and returns the final diagnostic.
+fn hint_invalid_main_file(
+    file_error: FileError,
+    input: FileId,
+) -> EcoVec<SourceDiagnostic> {
+    let is_utf8_error = matches!(file_error, FileError::InvalidUtf8);
+    let mut diagnostic =
+        SourceDiagnostic::error(Span::detached(), EcoString::from(file_error));
+
+    // Attempt to provide helpful hints for UTF-8 errors.
+    // Perhaps the user mistyped the filename.
+    // For example, they could have written "file.pdf" instead of
+    // "file.typ".
+    if is_utf8_error {
+        let path = input.vpath().as_rootless_path();
+        let extension = path.extension();
+
+        if let Some(extension) = extension {
+            if extension == "typ" {
+                // No hints if the file is already a .typ file.
+                // The file is indeed just invalid.
+                return eco_vec![diagnostic];
+            }
+
+            diagnostic.hint(eco_format!(
+                "a file with the `.{}` extension is not usually a Typst file",
+                extension.to_string_lossy()
+            ));
+        };
+
+        if path.with_extension("typ").exists() {
```

Comment:
Should we delete this checking?

Reply:
maybe we could try to fetch the FileId with the typ extension added from the world?

Reply:
The hint ```a file with the `.{extension}` extension is not usually a Typst file``` is already nice for me to guide the user to check directory by themself.

Reply:
cc @PgBiel

Reply:
In principle checking in `typst-cli` seemed more appropriate, since it's possible to use Typst without any files at all, but the current solution actually isn't too bad if we consider that the Typst model already works with "files", even if virtual. I'd say it's open for future improvement, but I would consider the current solution satisfactory for the time being.